File: ios/Shared/HomePolicy.swift
import Foundation

struct HomeWindow: Codable, Equatable { var start: String; var end: String }
struct HomeDayRule: Codable, Equatable { var days: [Int]; var allowance_minutes: Int; var windows: [HomeWindow] }
struct HomePolicy: Codable, Equatable {
	var timezone: String
	var away_usage_counts: Bool
	var outside_windows: String
	var rules: [HomeDayRule]
	var calendar: Calendar { var value = Calendar(identifier: .gregorian); value.timeZone = TimeZone(identifier: timezone)!; return value }
	func validate() throws {
		guard timezone == "America/Los_Angeles", !away_usage_counts, ["unrestricted", "block_at_home"].contains(outside_windows), (1...7).contains(rules.count) else { throw DocumentError.invalid("Unsupported home allowance settings.") }
		var days = Set<Int>()
		for rule in rules {
			guard (1...180).contains(rule.allowance_minutes), (1...2).contains(rule.windows.count), !rule.days.isEmpty else { throw DocumentError.invalid("Invalid daily allowance.") }
			for day in rule.days { guard (1...7).contains(day), days.insert(day).inserted else { throw DocumentError.invalid("Each weekday needs exactly one allowance.") } }
			var previous_end = 0
			for window in rule.windows {
				guard let start = ScheduleWindow.minutes(window.start), let end = ScheduleWindow.minutes(window.end), start >= previous_end, end - start >= 15 else { throw DocumentError.invalid("Home windows must be ordered, nonoverlapping, and at least 15 minutes.") }
				previous_end = end
			}
		}
		guard days.count == 7 else { throw DocumentError.invalid("Configure all seven weekdays.") }
	}
	// Outside the windows nothing is blocked unless the document kept the older setting on purpose.
	var blocks_outside: Bool { outside_windows == "block_at_home" }
	func rule(at date: Date) -> HomeDayRule? { rules.first { $0.days.contains(calendar.component(.weekday, from: date)) } }
	func allows(at date: Date) -> Bool {
		let minute = calendar.component(.hour, from: date) * 60 + calendar.component(.minute, from: date)
		return rule(at: date)?.windows.contains { minute >= (ScheduleWindow.minutes($0.start) ?? 1440) && minute < (ScheduleWindow.minutes($0.end) ?? 0) } == true
	}
	func day_key(_ date: Date) -> String { let c = calendar.dateComponents([.year, .month, .day], from: date); return "\(c.year!)-\(c.month!)-\(c.day!)" }
}

struct HomeLedger: Codable, Equatable {
	var day = ""
	var used_minutes = 0
	var segment_base = 0
	var generation: String?
	var bonuses: [String: Int]?
	// Past days' used/budget, keyed by day_key, so status reports can show a trend. Trimmed to StatusReport.history_days entries.
	var history: [String: [Int]]?
	mutating func reset_if_needed(policy: HomePolicy, now: Date) {
		let today = policy.day_key(now)
		if day != today {
			if !day.isEmpty {
				var past = history ?? [:]
				past[day] = [used_minutes, budget(policy.rules.first { rule in rule.days.contains(HomeLedger.weekday(of: day, policy: policy)) }?.allowance_minutes ?? 0)]
				if past.count > StatusReport.history_days { for key in past.keys.sorted().prefix(past.count - StatusReport.history_days) { past.removeValue(forKey: key) } }
				history = past
			}
			day = today; used_minutes = 0; segment_base = 0; generation = nil; bonuses = [:]
		}
	}
	// day_key is "YYYY-M-D"; recover its weekday so yesterday's budget is the right rule's.
	static func weekday(of key: String, policy: HomePolicy) -> Int {
		let parts = key.split(separator: "-").compactMap { Int($0) }
		guard parts.count == 3, let date = policy.calendar.date(from: DateComponents(year: parts[0], month: parts[1], day: parts[2])) else { return 0 }
		return policy.calendar.component(.weekday, from: date)
	}
	func budget(_ base: Int) -> Int { min(1440, base + (bonuses ?? [:]).values.reduce(0,+)) }
	mutating func grant(_ key: String, minutes: Int) throws -> Bool {
		guard (1...1440).contains(minutes) else { throw DocumentError.invalid("Choose 1 to 1440 bonus minutes.") }
		if bonuses?[key] != nil { return false }
		if bonuses == nil { bonuses = [:] }; bonuses?[key] = minutes; pause(); return true
	}
	mutating func checkpoint(generation: String, minutes: Int, at_home: Bool) {
		guard at_home, self.generation == generation, minutes > 0 else { return }
		used_minutes = max(used_minutes, segment_base + minutes)
	}
	mutating func pause() { generation = nil; segment_base = used_minutes }
}

File: lib/home-policy.ts
import { z } from "zod";

const time = z.string().regex(/^([01]\d|2[0-3]):[0-5]\d$/);
export const home_policy_schema = z.object({
	timezone: z.literal("America/Los_Angeles"),
	away_usage_counts: z.literal(false),
	// "unrestricted": outside the windows nothing is blocked (the default). "block_at_home": the older behavior, kept only for documents that chose it on purpose.
	outside_windows: z.enum(["unrestricted", "block_at_home"]),
	rules: z.array(z.object({
		days: z.array(z.number().int().min(1).max(7)).min(1).max(7),
		allowance_minutes: z.number().int().min(1).max(180),
		windows: z.array(z.object({ start: time, end: time }).strict()).min(1).max(2),
	}).strict()).min(1).max(7),
}).strict().superRefine((policy, context) => {
	const days = new Set<number>();
	for (const rule of policy.rules) {
		for (const day of rule.days) { if (days.has(day)) { context.addIssue({ code: "custom", message: "Each day has one shared allowance." }); } days.add(day); }
		let end = "00:00";
		for (const window of rule.windows) {
			if (window.start < end || window.end <= window.start || minutes(window.end) - minutes(window.start) < 15) { context.addIssue({ code: "custom", message: "Windows must be ordered, non-overlapping, and at least 15 minutes long within one day." }); }
			end = window.end;
		}
	}
	if (days.size !== 7) { context.addIssue({ code: "custom", message: "Choose an allowance for all seven days." }); }
});
export type HomePolicy = z.infer<typeof home_policy_schema>;
function minutes(time: string): number { const [h, m] = time.split(":").map(Number); return h * 60 + m; }
export const requested_home_policy: HomePolicy = {
	timezone: "America/Los_Angeles", away_usage_counts: false, outside_windows: "unrestricted",
	rules: [
		{ days: [2, 3, 4, 5], allowance_minutes: 30, windows: [{ start: "18:00", end: "18:30" }, { start: "19:00", end: "20:50" }] },
		{ days: [6], allowance_minutes: 120, windows: [{ start: "14:30", end: "20:20" }] },
		{ days: [1, 7], allowance_minutes: 180, windows: [{ start: "06:30", end: "20:30" }] },
	],
};
export function home_decision(policy: HomePolicy, day: number, time: string, at_home: boolean, consumed: number) {
	const rule = policy.rules.find((rule) => rule.days.includes(day));
	const allowed = Boolean(rule?.windows.some((window) => window.start <= time && time < window.end));
	const exhausted = consumed >= (rule?.allowance_minutes ?? 0);
	// Blocking happens inside a window once the minutes are spent; outside a window only the legacy setting blocks.
	const blocked = at_home && (allowed ? exhausted : policy.outside_windows === "block_at_home");
	return { count_usage: at_home && allowed && !exhausted, blocked, remaining: Math.max(0, (rule?.allowance_minutes ?? 0) - consumed) };
}
